sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.cosmos.spark
+
+import com.azure.cosmos.models.PartitionKey
+import com.azure.cosmos.{BulkOperations, CosmosAsyncContainer, CosmosBulkOperationResponse, CosmosItemOperation}
+import com.fasterxml.jackson.databind.node.ObjectNode
+import reactor.core.Disposable
+import reactor.core.publisher.EmitterProcessor
+import reactor.core.scala.publisher.SFlux
+import reactor.core.scala.publisher.SMono.PimpJFlux
+
+import java.util.concurrent.atomic.{AtomicInteger, AtomicReference}
+import java.util.concurrent.locks.ReentrantLock
+
+case class BulkWriter(container: CosmosAsyncContainer,
+                      writeConfig: CosmosWriteConfig) {
+
+  private val activeTasks = new AtomicInteger(0)
+  private val lock = new ReentrantLock
+  private val pendingTasksCompleted = lock.newCondition
+  private val errorCaptureFirstException = new AtomicReference[Throwable]()
+  private val bulkInputEmitter: EmitterProcessor[CosmosItemOperation] = EmitterProcessor.create
+
+  // TODO: moderakh implement retry
+  // TODO: moderakh handle 409
+
+  private val subscriptionDisposable: Disposable = {
+    val bulkOperationResponseFlux: SFlux[CosmosBulkOperationResponse[Object]] =
+      container.processBulkOperations[Object](bulkInputEmitter).asScala
+
+    bulkOperationResponseFlux.subscribe(
+      resp => {
+        resp.getBatchContext
+        if (resp.getException != null) {
+          // TODO: moderakh we should retry
+          captureIfFirstFailure(resp.getException)
```

Comment:
Would it be worth adding the concept of "retry policies"? For example, have one retry policy for 429, another retry policy for timeouts, and be able to compose them, so for a read flow you could just use the 429 + timeout but for a write flow, just the 429 one.
A retry policy would simple define the condition upon which an operation should retry for a particular scenario and can be reused?

Reply:
Also, couldn't bulk have different responses for each operation? So some operations might have succeeded and others failed and only those that failed could be retried?